Refine the Array.prototype.reduce() doc for clarity

sideshowbarker opened the pull request:
We have multiple open issues — #6156 and #6497 — reporting that the current description we have in MDN for how `Array.prototype.reduce()` works is confusing, and asking that we refine it to make it more clear.

So this change replaces the current description by citing the https://tc39.es/ecma262/#sec-array.prototype.reduce description in the spec — which is much more clear — and then updating related text and variable names to match the spec language.

One key difference that helps make the spec language much more clear is that it doesn’t use the word “accumulator” but instead uses “previousValue”. So for consistency, this change also replaces “accumulator” throughout with “previousValue".

Fixes https://github.com/mdn/content/issues/6156
Fixes https://github.com/mdn/content/issues/6497

The change also drops the entire Polyfill section — in keeping with our current policy of not any longer maintaining polyfill code in MDN articles, but instead linking to polyfills at external sites (and in this case, the article already has a link to a high-quality polyfill).

sideshowbarker committed:
Refine the Array.prototype.reduce() doc for clarity

We have multiple open issues — #6156 and #6497 — reporting that the
current description we have in MDN for how Array.prototype.reduce()
works is confusing, and asking that we refine it to make it more clear.

So this change replaces the current description by citing the
https://tc39.es/ecma262/#sec-array.prototype.reduce description in the
spec — which is much more clear — and then updating related text and
variable names to match the spec language.

One key difference that helps make the spec language much more clear is
that it doesn’t use the word “accumulator” but instead uses
“previousValue”. So for consistency, this change also replaces
“accumulator” throughout with “previousValue".

Fixes https://github.com/mdn/content/issues/6156
Fixes https://github.com/mdn/content/issues/6497

The change also drops the entire Polyfill section — in keeping with our
current policy of not any longer maintaining polyfill code in MDN
articles, but instead linking to polyfills at external sites (and in
this case, the article already has a link to a high-quality polyfill).

hamishwillee commented on files/en-us/web/javascript/reference/global_objects/array/reduce/index.md:
```diff
@@ -15,112 +15,81 @@ browser-compat: javascript.builtins.Array.reduce
-Your **reducer** function's returned value is assigned to the accumulator,
-whose value is remembered across each iteration throughout the array, and ultimately
-becomes the final, single resulting value.
+{{EmbedInteractiveExample("pages/js/array-reduce.html")}}
```
FYI This still uses the old accumulator terminology. Modify for consistency?

sideshowbarker replied:
> FYI This still uses the old accumulator terminology. Modify for consistency?

Yes, thanks for reminding me — I plan on updating that interactive example after this PR is merged (but would rather not block this PR on that, because it’s a chicken/egg race condition; one or the other would need to land first, regardless).

sideshowbarker replied:
> Yes, thanks for reminding me — I plan on updating that interactive example after this PR is merged

Opened https://github.com/mdn/interactive-examples/pull/1887

hamishwillee commented on files/en-us/web/javascript/reference/global_objects/array/reduce/index.md:
```diff
@@ -15,112 +15,81 @@ browser-compat: javascript.builtins.Array.reduce
-The **`reduce()`** method executes a **reducer**
-function (that you provide) on each element of the array, resulting in a single output
-value.
+The **`reduce()`** method executes a “reducer” callback function (that you provide) on each element of the array, resulting in a single output value (the result of running the reducer to completion over the entire array).
```
```suggestion
The **`reduce()`** method executes a user-supplied “reducer” callback function on each element of the array, passing in the return value from the calculation on the preceding element. The final result of running the reducer across all elements of the array is a single value. 
```

I prefer the original text - the extra words don't help. What I'm trying to capture in this suggestion is that this was missing the concept that the results of the each call reflect the previous call. That's accumulation :-)

hamishwillee replied:
PS I don't love my words so not precious if you want to consider alternatives.

sideshowbarker replied:
> PS I don't love my words so not precious if you want to consider alternatives.

I quite like your words :) And I agree it’s an improvement — it reads a lot more smoothly, without the double-parentheticals.

sideshowbarker committed:
Update files/en-us/web/javascript/reference/global_objects/array/reduce/index.md

Co-authored-by: Hamish Willee <hamishwillee@gmail.com>

hamishwillee commented on files/en-us/web/javascript/reference/global_objects/array/reduce/index.md:
```diff
@@ -15,112 +15,81 @@ browser-compat: javascript.builtins.Array.reduce
-2.  Current Value
-3.  Current Index
-4.  Source Array
+Perhaps the easiest-to-understand case for `reduce()` is to return the sum of all the elements in an array, by walking through the array element-by-element and accumulating a running sum of all the elements that precede each element, and then adding each element in turn to that accumulated sum — until there are no more elements to add — as shown in the following interactive example:
```
This sentence is too long/hard to parse. I'm having trouble rewriting it but minimally perhaps a full stop 
```suggestion
Perhaps the easiest-to-understand case for `reduce()` is to return the sum of all the elements in an array. 
The reducer walks through the array element-by-element, at each step adding the current array value to the result from the previous step (this result is the running sum of all the previous steps) — until there are no more elements to add.
This is shown in the following interactive example:
```

sideshowbarker replied:
Yes. This is better for clarity and for breaking up the paragraph length for better readability.

sideshowbarker committed:
Update files/en-us/web/javascript/reference/global_objects/array/reduce/index.md

Co-authored-by: Hamish Willee <hamishwillee@gmail.com>

hamishwillee commented on files/en-us/web/javascript/reference/global_objects/array/reduce/index.md:
```diff
@@ -15,112 +15,81 @@ browser-compat: javascript.builtins.Array.reduce
-    `currentValue`. Calling `reduce()` on an empty array
-    without an `initialValue` will throw a {{jsxref("TypeError")}}.
+- *callbackFn*
+  - : A “reducer” function that takes four arguments:
```
FWIW Don't know if we're supposed to use bullets or nested definition lists. I don't think it matters here.

hamishwillee replied:
@sideshowbarker But everywhere else parameters are in code style - e.g. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter

sideshowbarker replied:
> FWIW Don't know if we're supposed to use bullets or nested definition lists. I don't think it matters here.

Yeah it was definition list to start with, so just kept it that way.

sideshowbarker replied:
> @sideshowbarker But everywhere else parameters are in code style - e.g. [developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter)

Yeah, I don’t like that style, but I’ll change it here for consistency (and push the change in a minute)

hamishwillee commented on files/en-us/web/javascript/reference/global_objects/array/reduce/index.md:
```diff
@@ -15,112 +15,81 @@ browser-compat: javascript.builtins.Array.reduce
+    - *array* (the array to traverse) {{optional_inline}}
 
-### Return value
+- *initialValue* {{optional_inline}}
```
Ditto - should this be code style (and inline below)

hamishwillee commented on files/en-us/web/javascript/reference/global_objects/array/reduce/index.md:
```diff
@@ -15,112 +15,83 @@ browser-compat: javascript.builtins.Array.reduce
 If the array only has one element (regardless of position) and no
-`initialValue` is provided, or if
-`initialValue` is provided but the array is empty, the solo value
+*initialValue* is provided, or if
```
And I see you are being consistent within this doc.

sideshowbarker replied:
> And I see you are being consistent within this doc.

😆

hamishwillee commented on files/en-us/web/javascript/reference/global_objects/array/reduce/index.md:
```diff
@@ -15,112 +15,83 @@ browser-compat: javascript.builtins.Array.reduce
 
-If the array is empty and no `initialValue` is provided,
-{{jsxref("TypeError")}} will be thrown.
+The ECMASCript spec describes the behavior of `reduce()` as follows:
```
ECMASCript  or ECMAScript ?

sideshowbarker replied:
> ECMASCript or ECMAScript ?

oofs, thanks — fixed

hamishwillee reviewed (approved):
So much better. Minor comments but is approved. Feel free to self merge if you want, or you can ping me and I will do it.

sideshowbarker committed:
Fixes from review comments

Also, for unordered-list-marker consistency, use `-` rather than `*`,
and add a blank line before the list (to make markdownlint happy).

sideshowbarker commented:
> So much better. Minor comments but is approved. Feel free to self merge if you want, or you can ping me and I will do it.

Please if you have time, give it a once-over to make sure I didn’t regress anything when I pushed the latest commit. And if looks right, please do go ahead and merge.

hamishwillee commented:
@sideshowbarker Thank you for your 1E45 th contribution to MDN :-)
